core/trino-spi/src/main/java/io/trino/spi/type/TypeUtils.java

```diff
@@ -73,10 +73,18 @@ else if (type.getJavaType() == boolean.class) {
             type.writeBoolean(blockBuilder, (Boolean) value);
         }
         else if (type.getJavaType() == double.class) {
-            type.writeDouble(blockBuilder, ((Number) value).doubleValue());
+            if(value instanceof Float) {
+                type.writeDouble(blockBuilder, ((Float) value).doubleValue());
+            } else {
+                type.writeDouble(blockBuilder, ((Double) value).doubleValue());
```

Comment:
if we were writing this code anew, we  would require value to be Double.

what would it take to have that?

Reply:
to be clear: changing from "Number" to "Float or Double" is an important improvement

Reply:
Hi @findepi, I didn't understand what I need to change. There was an error in the CI related to the cast of an Integer as an Long, so I verified if the instance of the value was Integer and made the cast using longValue(). Same to Double and Float.

Reply:
> There was an error in the CI related to the cast of an Integer as an Long,

do you remember what was failing?

Reply:
> > There was an error in the CI related to the cast of an Integer as an Long,
> 
> do you remember what was failing?

java.lang.ClassCastException: class java.lang.Integer cannot be cast to class java.lang.Long (java.lang.Integer and java.lang.Long are in module java.base of loader 'bootstrap')

The same is appearing now with Short and Long.

Reply:
I think there are a lot of callers which send wider types than required. Those will need to be changed to just send Double instead of Float or something else.

If you can try to figure out how many callers you need to fix then we can decide to this now or as a follow-up. I would prefer now unless it seems like a very large change.

Reply:
> I think there are a lot of callers which send wider types than required. Those will need to be changed to just send Double instead of Float or something else.
> 
> If you can try to figure out how many callers you need to fix then we can decide to this now or as a follow-up. I would prefer now unless it seems like a very large change.

I couldn't find out, there are many callers for this function and for the calling functions. I thought of the following:
```suggestion
public static void writeNativeValue(Type type, BlockBuilder blockBuilder, @Nullable Object value)
    {
        ...
        else if (type.getJavaType() == double.class && value instanceof Double) {
            type.writeDouble(blockBuilder, ((Double) value).doubleValue());
        }
        else if (type.getJavaType() == long.class && value instanceof Long) {
            type.writeLong(blockBuilder, ((Long) value).longValue());
        }
        else if (type.getJavaType() == Slice.class) {
        ...
        else {
            type.writeObject(blockBuilder, value);
        }
    }
```

Would that be a good resolution, even a temporary one?